addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.unifi.internal.api;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.openhab.binding.unifi.internal.api.model.UniFiClient;
+import org.openhab.binding.unifi.internal.api.model.UniFiDevice;
+import org.openhab.binding.unifi.internal.api.model.UniFiSite;
+import org.openhab.binding.unifi.internal.api.util.UniFiClientDeserializer;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The {@link UniFiController} is the main communication point with an external instance of the Ubiquiti Networks
+ * Controller Software.
+ *
+ * @author Matthew Bowman - Initial contribution
+ */
+@NonNullByDefault
+public class UniFiController {
+
+    private HttpClient httpClient;
+
+    private String host;
+
+    private int port;
+
+    private String username;
+
+    private String password;
+
+    private Gson gson;
+
+    public UniFiController(String host, int port, String username, String password) {
+        this.host = host;
+        this.port = port;
+        this.username = username;
+        this.password = password;
+        this.httpClient = new HttpClient(new SslContextFactory(true)); // mgb: true = trust all
```

Comment:
You should not instantiate your own HttpClient, but rather use the `HttpClientFactory` from ESH. You'll have to get it injected in your thing handler factory and pass it through the constructor of the handler down to here where it's needed.

Reply:
IIRC the issue with using `HttpClientFactory` was I couldn't use a custom `SslContext` which was necessary as the UniFi controller ships with a self-signed certificate. 

If there's another way to override the SSL verification using an injected `HttpClient` instance, I can refactor.

Reply:
This is the place where the [TlsCertificateProvider](https://github.com/eclipse/smarthome/pull/6281/files#diff-1719561fd7d5b8ac889a486c612005a1R27) could shine.

The SHA2 fingerprint of the certificate used by my controller is `79 F3 F9 6C 6D 7A F3 17 85 B7 9D CD 0E B2 0C 50 62 C2 4F 1B 17 56 20 29 B6 2E 5F 70 00 4D D6 12` for controller software version `5.9.29`.

If this does not work, then the alternative could be to implement a [TlsTrustManagerProvider](https://github.com/eclipse/smarthome/pull/6281/files#diff-bf7828c630dea0acfe9071f9ed62a5faR27). Which has not yet been done before :-)

For the implementation you can use 'UniFi' for `String getHostName();` because that is what unifi set as a common name for the certificate.

Reply:
I don't think the `TlsCertificateProvider` will work as I believe everybody's certificate will be different.

That being said, maybe a `TlsTrustManagerProvider` that "accepts all" for the `config.getHostname()` parameter would work.

I'm trying to implement this, but I'm unsure of the exact "wiring" I need.

Using this in `UniFiThingHandlerFactory` I can get (and pass through) the common `HttpClient`:

```
@Reference
public void setHttpClientFactory(HttpClientFactory httpClientFactory) {
    this.httpClient = httpClientFactory.getCommonHttpClient();
}
```

However I'm not sure where / how to "install" the `TlsTrustManagerProvider`. I've tried this but it doesn't seem to be right...

```
if (UniFiControllerThingHandler.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
    handler = new UniFiControllerThingHandler((Bridge) thing, httpClient);
    bundleContext.registerService(TlsTrustManagerProvider.class.getName(), handler, null);
}
```

with `UniFiControllerThingHandler` implementing `TlsTrustManagerProvider`.

Reply:
> Maybe, maybe not? That is why I gave you my SHA2 fingerprint :-)

I have three hosts running three different versions all with different fingerprints ;-)

I manage to switch to the common `HttpClient` instance as well as implement the `TlsTrustManagerProvider`

I need to fix my Jenkins instance and ask the forum to test this since it's a rather important / critical bit of the binding that changed.